docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+               $data['address'] = $this->addressCodec->encode($value->address);
+           }
+
+           return MongoDB\BSON\Document::fromPHP($data);
+       }
+   }
+
+Codec Libraries
+---------------
+
+If you have a number of codecs that you want to use in multiple places, you can create a codec library. A codec library
+contains a list of codecs and checks each codec if it supports a value. If it does, it will use that codec to encode or
+decode the given value. The following code snippet changes the ``PersonCodec`` to use a codec library instead of a
+hard-coded ``DateTimeCodec``:
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       public function __construct(
+           private readonly MongoDB\Codec\CodecLibrary $codecLibrary,
+       ) {}
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+           $person->createdAt = $this->codecLibrary->decode($value->get('createdAt'));
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+               'createdAt' => $this->codecLibrary->encode($value->createdAt),
+           ]);
+       }
+   }
+
+This way, you can configure a global codec library and reuse it in multiple places:
+
+.. code-block:: php
+
+   <?php
+
+    $codecLibrary = new MongoDB\Codec\CodecLibrary([
+         new DateTimeCodec(),
+    ]);
+
+    $personCodec = new PersonCodec($codecLibrary);
+
+.. note::
+
+   Be careful when adding ``DocumentCodec`` instances to a codec library. The codec library will only use the first
+   codec that supports a given value. While this will most likely work fine when encoding to BSON, it can lead to issues
+   when decoding from BSON as you'll have to inspect the BSON document to detect whether the codec supports it. For this
+   reason, the ``codec`` option in a collection does not accept a codec library.
+
+Library-aware Codecs
+~~~~~~~~~~~~~~~~~~~~
+
+If you have a codec that needs to be aware of the codec library, you can implement the
+``MongoDB\Codec\KnowsCodecLibrary`` interface. When the codec is added to a library, this library is then injected into
+the codec. This allows you to create a common library for all value codecs and re-use it in multiple document codecs.
+
+.. note::
+
+   The ``attachCodecLibrary`` method will not be called if a library-aware codec is never added to a library.
```

Comment:
Is there an edge case where a codec is added to multiple libraries? In that case, `attachCodecLibrary()` may get called multiple times, no?

I realize KnowsCodecLibrary is only an interface and we don't ship a trait, so this is up to the user to implement. I'm wondering if they might shoot themselves in the foot with this, though.

Reply:
That is correct, adding the same codec instance to multiple libraries may be an issue, but that shouldn't happen too often. Let me know if you think this is worth calling out in the docs.